Ensure QDatePicker sets date ranges and dates correctly.

chris2oph opened the pull request:
This change fixes the test failures for the **DatePickerCtrlTestCase**. The problem here is that the default `QDate`s were being set as: **25/11/-4714** to **31/12/7999**. The test **Range** was checking if the end date was invalid, and the default end date for QT was considered to be valid (which fails the test). This necessitated the inclusion of the `IsQDateValid` function; I do not particularly like the usage of the constant to represent the max year for QT however I tried a lot of different things and this seemed to be the lesser of all evils for me in this instance.

The inclusion of the extra check in the `SetDate` function is necessary to ensure that any date being set must fall between the specified range.

Any suggestions or comments on this change would be welcomed.

chris2oph committed:
Added validation on date ranges for QDates. Ensured that default max date is considered invalid. Also ensured that any attempts to set a date value check that the proposed date is within the range.

catalinr commented on src/qt/calctrl.cpp:
```diff
@@ -168,15 +180,22 @@ bool wxCalendarCtrl::SetDateRange(const wxDateTime& lowerdate,
+        status = false;
 
-    if (lowerdate)
+    if ( lowerdate && IsQDateValid( m_qtCalendar->minimumDate() ) )
```
How will this not crash when `m_qtCalendar` is null but you no longer `return` in that case, 2 lines above?

chris2oph replied:
A very good question, I will do some testing around this and get back to you!

chris2oph replied:
In short... it will crash! Do you think the best idea would be to remove the boolean `status` flag altogether and simply have multiple return statements or to keep the `status` flag and change line 188 to: `return false`?

catalinr replied:
I don't know the code, but the best thing is almost always to start with the minimal changes that are needed to achieve your goal (see the removal of an empty line that could probably have been avoided too).

chris2oph replied:
I have made a further change which should hopefully address these issues. Many thanks for your input @catalinr

chris2oph committed:
Reinstated empty line that was accidentally removed and also changed a boolean return flag to a return statement to prevent crash.

chris2oph committed:
Corrected return value logic of GetDateRange() function.

chris2oph commented on src/qt/calctrl.cpp:
```diff
@@ -168,15 +181,22 @@ bool wxCalendarCtrl::SetDateRange(const wxDateTime& lowerdate,
         *lowerdate = wxQtConvertDate(m_qtCalendar->minimumDate());
-    if (upperdate)
+    else
+        status = false;
```
You are correct on both counts, I have updated with another change to address these. Many Thanks!

vadz commented on include/wx/qt/calctrl.h:
```diff
@@ -78,6 +79,7 @@ class WXDLLIMPEXP_ADV wxCalendarCtrl : public wxCalendarCtrlBase
 private:
     void Init();
     void UpdateStyle();
+    bool IsQDateValid(const QDate &date) const;
```
I think this function is an implementation detail, so I'd keep it in `src/qt/calctrl.cpp` instead of making it a member -- it doesn't really need to be a member function, and so shouldn't be one.

vadz commented on src/qt/calctrl.cpp:
```diff
@@ -24,6 +24,11 @@
 
+namespace
+{
+    const int MAX_YEAR_QT = 7999;
```
I'd add a comment saying that the native control doesn't support years beyond this one (at least if this is true, i.e. if I understand correctly what's going on here).

vadz commented on src/qt/calctrl.cpp:
```diff
@@ -120,6 +125,11 @@ void wxCalendarCtrl::UpdateStyle()
 
+bool wxCalendarCtrl::IsQDateValid(const QDate &date) const
+{
+    return date.isValid() && !date.isNull() && date.day() > 0 && date.month() > 0 && date.year() > 0 && date.year() < MAX_YEAR_QT;
```
`isValid()` is the same as `isNull()` according to Qt docs, so just one of those is needed. And I'm pretty sure that a valid date can't have neither day nor the month of 0. OTOH I'm not sure why does it need to have a positive year (what if want the user to enter Aristotle's birth date, eh?).

So, all in all, I'd change this line like this:
```suggestion
    return date.isValid() && date.year() < MAX_YEAR_QT;
```

chris2oph replied:
The suggestion to change the check here so that: `date.year() > 0` is removed causes the test to fail with the following exception during the call `wxQtConvertDate()`: _date out of range - can't convert to JDN_

I appreciate your suggestion to remove this check in terms of reasonable usage of date pickers, however I believe we will need to keep this. All of your other suggested improvements are valid and are appreciated.

All of this may not be necessary if we instead add `#ifndef __WXQT__` around the corresponding unit test and mentioning that Qt port limits the entry to the dates before 8000-01-01 in the documentation. I will make the suggested changes and add a further commit shortly, then we can make a decision on whether or not we keep these changes or simply do not perform these tests for QT.

vadz reviewed (commented):
I'm not actually that certain that we should be doing this at all and I could settle with adding `#ifndef __WXQT__` around the corresponding unit test and mentioning that Qt port limits the entry to the dates before 8000-01-01 in the documentation.

But if you prefer to do it like this, I don't see anything really wrong with it neither, although arguably the docs still should mention that dates after the max one are not supported.

chris2oph committed:
Made QDate validation function a local function rather than a member and also removed some superfluous checks in this function.

chris2oph committed:
Removed custom date validity checks for QDates and ensured end date validity check is omitted for QDates in Range test for DatePickerCtrlTest.

chris2oph commented:
I have addressed @vadz latest comment and have removed the `IsQDateValid()` function. I actually agree that we would be better off simply ignoring this test for QT and so I have added the `ifndef __wxQT__` around the failing assertion for this test. I hope this resolves this issue, however if there are any issues please let me know.

vadz commented:
Thanks, I've (finally) applied this with a couple of cosmetic changes.
